combine_chunks.py

```diff
@@ -90,15 +102,83 @@ def combine_chunks(
+            code.attrs["git_hash"] = np.bytes_(args.git_hash.encode('utf-8'))
+
+            # Copy swift metadata
+            params = cellgrid.copy_swift_metadata(outfile)
+
+            # Generate header
+            header = outfile.create_group('Header')
+            for attr in [
+                    'BoxSize',
+                    'Dimension',
+                    'NumPartTypes',
+                    'Redshift',
+                    'RunName',
+                    'Scale-factor',
+                ]:
+                header.attrs[attr] = cellgrid.swift_header_group[attr]
+            header.attrs['Code'] = np.bytes_('SOAP'.encode('utf-8'))
+            header.attrs['Dimension'] = cellgrid.swift_header_group['Dimension']
+            header.attrs['NumFilesPerSnapshot'] = np.array([1], dtype='int32')
+            header.attrs['NumSubhalos_ThisFile'] = np.array([total_nr_halos], dtype='int32')
+            header.attrs['NumSubhalos_Total'] = np.array([total_nr_halos], dtype='int32')
+            n_part_type = cellgrid.swift_header_group['NumPartTypes'][0]
+            header.attrs['NumPart_ThisFile'] = np.zeros(n_part_type, dtype='int32')
+            header.attrs['NumPart_Total'] = np.zeros(n_part_type, dtype='uint32')
+            header.attrs['NumPart_Total_Highword'] = np.zeros(n_part_type, dtype='uint32')
+            header.attrs['OutputType'] = np.bytes_('SOAP'.encode('utf-8'))
+            snapshot_date = time.strftime("%H:%M:%S %Y-%m-%d GMT", time.gmtime())
+            header.attrs['SnapshotDate'] = np.bytes_(snapshot_date.encode('utf-8'))
+            # TODO:
+            header.attrs['System'] = np.bytes_(socket.gethostname().encode('utf-8'))
+            header.attrs['ThisFile'] = np.array([0], dtype='int32')
+
+            # Write cosmology
+            cosmo = outfile.create_group("Cosmology")
+            for name, value in cellgrid.cosmology.items():
+                cosmo.attrs[name] = [value]
+
+            # Write units
+            units = outfile.create_group("Units")
+            for name, value in cellgrid.swift_units_group.items():
+                units.attrs[name] = [value]
+            # TODO: Is this correct?
+            units.attrs['Unit mass in cgs (U_M)'] = [unyt.solar_mass.to('g')]
+
+            # Write physical constants
+            const = outfile.create_group("PhysicalConstants")
+            const = const.create_group("CGS")
+            for name, value in cellgrid.constants.items():
+                const.attrs[name] = [value]
+
+            # Write cell information
+            cells = outfile.create_group("Cells")
+            cells_metadata = cells.create_group("Meta-data")
+            cells_metadata.attrs['dimension'] = cellgrid.dimension
+            cells_metadata.attrs['nr_cells'] = cellgrid.nr_cells
+            cell_size_cMpc = cellgrid.cell_size.to(cellgrid.boxsize.units).value
+            cells_metadata.attrs['size'] = cell_size_cMpc
```

Comment:
Is this right? I thought we saved this in the units of the file.

Reply:
I guess it's not guaranteed to be Mpc, but cellgrid.boxsize has the correct units, so I think this should be ok after I change the variable name.

Reply:
The ref_metadata parameter passed into this function contains unit information for all halo properties. We should look up the units there. If I run this branch in pdb then I can see that cellgrid.boxsize.units = a*snap_length but the units of cofp in ref_metadata are a*swift_mpc, which is not guaranteed to be the same.